authentication/provider/factory/linking_provider_factory.go

```diff
@@ -0,0 +1,99 @@
+package factory
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/fabric8-services/fabric8-auth/application/service"
+	"github.com/fabric8-services/fabric8-auth/application/service/base"
+	servicecontext "github.com/fabric8-services/fabric8-auth/application/service/context"
+	account "github.com/fabric8-services/fabric8-auth/authentication/account/repository"
+	"github.com/fabric8-services/fabric8-auth/authentication/provider"
+	errs "github.com/fabric8-services/fabric8-auth/errors"
+	"github.com/fabric8-services/fabric8-auth/log"
+	"github.com/satori/go.uuid"
+)
+
+// NewLinkingProviderFactory returns the default Linking provider factory.
+func NewLinkingProviderFactory(context servicecontext.ServiceContext, config provider.LinkingProviderConfiguration) service.LinkingProviderFactory {
+	factory := &linkingProviderFactoryImpl{
+		BaseService: base.NewBaseService(context),
+		config:      config,
+	}
+	return factory
+}
+
+type linkingProviderFactoryImpl struct {
+	base.BaseService
+	config provider.LinkingProviderConfiguration
+}
+
+// NewLinkingProvider creates a new linking provider for the given resource URL or provider alias
+func (f *linkingProviderFactoryImpl) NewLinkingProvider(ctx context.Context, identityID uuid.UUID, authURL string, forResource string) (provider.LinkingProvider, error) {
+	// Check if the forResource is actually a provider alias like "github" or "openshift"
+	if forResource == provider.GitHubProviderAlias {
+		return provider.NewGitHubIdentityProvider(f.config.GetGitHubClientID(), f.config.GetGitHubClientSecret(), f.config.GetGitHubClientDefaultScopes(), authURL), nil
+	}
+	if forResource == provider.OpenShiftProviderAlias {
+		// Look up the user's OpenShift cluster
+		var clusterURL string
+		err := f.ExecuteInTransaction(func() error {
+			identities, err := f.Repositories().Identities().Query(account.IdentityFilterByID(identityID), account.IdentityWithUser())
+			if err != nil {
+				return err
+			}
+			if len(identities) == 0 {
```

Comment:
same here, I believe this check is not necessary

Reply:
Isn't it necessary to check if no identities were found by the query?  Otherwise the very next check will result in an out of index error.